shell/platform/android/io/flutter/embedding/engine/dynamicfeatures/PlayStoreDynamicFeatureManager.java

```diff
@@ -0,0 +1,305 @@
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.dynamicfeatures;
+
+import android.content.Context;
+import android.content.pm.PackageManager.NameNotFoundException;
+import android.content.res.AssetManager;
+import android.os.Build;
+import androidx.annotation.NonNull;
+import com.google.android.play.core.splitinstall.SplitInstallException;
+import com.google.android.play.core.splitinstall.SplitInstallManager;
+import com.google.android.play.core.splitinstall.SplitInstallManagerFactory;
+import com.google.android.play.core.splitinstall.SplitInstallRequest;
+import com.google.android.play.core.splitinstall.SplitInstallSessionState;
+import com.google.android.play.core.splitinstall.SplitInstallStateUpdatedListener;
+import com.google.android.play.core.splitinstall.model.SplitInstallErrorCode;
+import com.google.android.play.core.splitinstall.model.SplitInstallSessionStatus;
+import io.flutter.Log;
+import io.flutter.embedding.engine.FlutterJNI;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+
+/**
+ * Flutter default implementation of DynamicFeatureManager that downloads dynamic feature modules
+ * from the Google Play store.
+ */
+public class PlayStoreDynamicFeatureManager implements DynamicFeatureManager {
+  private static final String TAG = "flutter";
+
+  private @NonNull SplitInstallManager splitInstallManager;
+  private @NonNull Map<Integer, String> sessionIdToName;
+  private @NonNull Map<Integer, Integer> sessionIdToLoadingUnitId;
+  private @NonNull FlutterJNI flutterJNI;
+  private @NonNull Context context;
+
+  private FeatureInstallStateUpdatedListener listener;
+
+  private class FeatureInstallStateUpdatedListener implements SplitInstallStateUpdatedListener {
+    public void onStateUpdate(SplitInstallSessionState state) {
+      if (sessionIdToName.containsKey(state.sessionId())) {
+        // TODO(garyq): Add capability to access the state from framework.
+        switch (state.status()) {
+          case SplitInstallSessionStatus.FAILED:
+            {
+              Log.e(
+                  TAG,
+                  String.format("Module \"%s\" (sessionId %d) install failed with: %s",
+                      sessionIdToName.get(state.sessionId()),
+                      state.sessionId(),
+                      state.errorCode())
+              );
+              flutterJNI.dynamicFeatureInstallFailure(
+                  sessionIdToName.get(state.sessionId()),
+                  sessionIdToLoadingUnitId.get(state.sessionId()),
+                  "Module install failed with " + state.errorCode(),
+                  true);
+              sessionIdToName.remove(state.sessionId());
+              sessionIdToLoadingUnitId.remove(state.sessionId());
+              break;
+            }
+          case SplitInstallSessionStatus.INSTALLED:
+            {
+              Log.e(
```

Comment:
Yeah, during dev, I change them to .e because they show up in surface level logging and is much easier to work with, They will become debug logs in this PR, and in a future PR, will probably be changed to some system to expose the state to the developer.

Reply:
Use https://github.com/flutter/engine/blob/c0f7e8acf9318d264ad6a235facd097de597ffcc/shell/platform/android/io/flutter/Log.java#L29 instead